solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -1,26 +1,131 @@
+use std::str::FromStr;
+
+use anchor_lang::prelude::{AccountInfo, CpiContext, Pubkey};
 use anchor_lang::solana_program::msg;
+use anchor_lang::AccountDeserialize;
+use anchor_spl::token::{spl_token, Burn, MintTo, TokenAccount, Transfer};
 use ibc::applications::transfer::context::{
     TokenTransferExecutionContext, TokenTransferValidationContext,
 };
 use ibc::applications::transfer::error::TokenTransferError;
-use ibc::applications::transfer::PrefixedCoin;
+use ibc::applications::transfer::{Amount, PrefixedCoin};
 use ibc::core::ics24_host::identifier::{ChannelId, PortId};
-use ibc::Signer;
+use primitive_types::U256;
+use strum::Display;
+use uint::FromDecStrErr;
 
-// use crate::module_holder::IbcStorage<'_,'_>;
+use crate::storage::ids::PortChannelPK;
 use crate::storage::IbcStorage;
+use crate::MINT_ESCROW_SEED;
+
+/// Structure to identify if the account is escrow or not. If it is escrow account, we derive the escrow account using port-id, channel-id and denom.
+#[derive(
+    Clone, Display, PartialEq, Eq, derive_more::From, derive_more::TryInto,
+)]
+pub enum AccountId {
+    Signer(Pubkey),
+    Escrow(PortChannelPK),
+}
+
+impl TryFrom<ibc::Signer> for AccountId {
+    type Error = <Pubkey as FromStr>::Err;
+
+    fn try_from(value: ibc::Signer) -> Result<Self, Self::Error> {
+        Ok(Pubkey::try_from(value.as_ref()).map(Self::Signer).unwrap())
+    }
+}
+
+impl AccountId {
+    pub fn get_escrow_account(&self, denom: &str) -> Result<Pubkey, ()> {
+        let port_channel = match self {
+            AccountId::Escrow(pk) => pk,
+            AccountId::Signer(_) => return Err(()),
+        };
+        let channel_id = port_channel.channel_id();
+        let port_id = port_channel.port_id();
+        let seeds =
+            [port_id.as_bytes(), channel_id.as_bytes(), denom.as_bytes()];
+        let (escrow_account_key, _bump) =
+            Pubkey::find_program_address(&seeds, &crate::ID);
+        Ok(escrow_account_key)
+    }
+}
 
-impl TokenTransferExecutionContext for IbcStorage<'_, '_> {
+impl TryFrom<&AccountId> for Pubkey {
+    type Error = ();
+
+    fn try_from(value: &AccountId) -> Result<Self, Self::Error> {
+        match value {
+            AccountId::Signer(signer) => Ok(*signer),
+            AccountId::Escrow(_) => Err(()),
+        }
+    }
```

Comment:
I think it makes sense to take advantage of existing TryFrom
implementation that derive_more gives us:

```suggestion
    type Error = <Self as TryFrom<AccountId>>::Error;;

    fn try_from(value: &AccountId) -> Result<Self, Self::Error> {
        Self::try_from(*value)
    }
```